diffyscan/utils/calldata.py

```diff
@@ -0,0 +1,45 @@
+from .logger import logger
+from .encoder import encode_constructor_arguments
+
+def get_calldata(contract_address_from_config, target_compiled_contract, binary_config):
+    calldata = get_prepared_calldata_from_config(contract_address_from_config, binary_config)
+    if calldata is not None:
+          return calldata, ''
+    
+    calldata, text_error = parse_calldata_from_config(
+        contract_address_from_config,
+        binary_config["constructor_args"],
+        target_compiled_contract,
+    )
+    return calldata, text_error
+            
+def get_constructor_abi(target_compiled_contract):
+    constructor_abi = None
+    try:
+        constructor_abi = [entry["inputs"] for entry in target_compiled_contract['abi'] if entry["type"] == "constructor"][0]
+    except IndexError:
+        logger.okay(f'Constructor in ABI not found, calldata calculation skipped')
+        return None
+
+    logger.okay(f'Constructor in ABI successfully found: {[arg['type'] for arg in constructor_abi]}')
+    return constructor_abi
+
+def get_prepared_calldata_from_config(contract_address_from_config, binary_config):
+    if "constructor_calldata" not in binary_config or contract_address_from_config not in binary_config["constructor_calldata"]:
+        return None
+    calldata_field = binary_config["constructor_calldata"]
+    logger.info(f"Trying to use prepared calldata from config")
+    prepared_calldata_from_config = calldata_field[contract_address_from_config]
+    return prepared_calldata_from_config
+    
+def parse_calldata_from_config(contract_address_from_config, constructor_args, target_compiled_contract):    
```

Comment:
If you return composed types, consider providing type hints for the function declarations, e.g.:

```python
def parse_calldata_from_config(
    contract_address_from_config, 
    constructor_args, 
    target_compiled_contract
) → (Optional[str], Optional[str]):
```

Reply:
Nevertheless, I am not sure that we should use Optional types and not raising exceptions

The file feels like a mix of both approaches, but I may miss an intended structure

Reply:
reworked with  exceptions usage